src/core/components/pin.js

```diff
@@ -0,0 +1,431 @@
+        hashes = null
+      }
+      if (typeof options === 'function') {
+        callback = options
+      }
+      if (hashes && hashes.type) {
+        options = hashes
+        hashes = null
+      }
+      if (options && options.type) {
+        type = options.type.toLowerCase()
+      }
+      callback = once(callback)
+      if (Object.keys(pin.types).indexOf(type) < 0) {
+        return callback(new Error(
+          `Invalid type '${type}', must be one of {direct, indirect, recursive, all}`
+        ))
+      }
+      if (hashes) {
+        normalizeHashes(self, hashes, (err, mhs) => {
+          if (err) { return callback(err) }
+          const result = {
+            // async result queue
+            payload: [],
+            update: (item) => {
+              result.payload.push(item)
+              if (result.payload.length === mhs.length) {
+                return callback(null, result.payload)
+              }
+            }
+          }
+          mhs.forEach((multihash) => {
+            pin.isPinnedWithType(multihash, type, (err, pinned, reason) => {
+              if (err) { return callback(err) }
+              const key = toB58String(multihash)
+              if (!pinned) {
+                return callback(new Error(
+                  `Path ${key} is not pinned`
+                ))
+              }
+              switch (reason) {
+                case pin.types.direct:
+                case pin.types.recursive:
+                  result.update({
+                    hash: key,
+                    type: reason
+                  })
+                  break
+                default:
+                  result.update({
+                    hash: key,
+                    type: `${pin.types.indirect} through ${reason}`
+                  })
+              }
+            })
+          })
+        })
+      } else {
+        const result = []
+        if (type === pin.types.direct || type === pin.types.all) {
+          pin.directKeyStrings().forEach((hash) => {
+            result.push({
+              type: pin.types.direct,
+              hash: hash
+            })
+          })
+        }
+        if (type === pin.types.recursive || type === pin.types.all) {
+          pin.recursiveKeyStrings().forEach((hash) => {
+            result.push({
+              type: pin.types.recursive,
+              hash: hash
+            })
+          })
+        }
+        if (type === pin.types.indirect || type === pin.types.all) {
+          pin.getIndirectKeys((err, hashes) => {
+            if (err) { return callback(err) }
+            hashes.forEach((hash) => {
+              result.push({
+                type: pin.types.indirect,
+                hash: hash
+              })
+            })
+            return callback(null, result)
+          })
+        } else {
+          return callback(null, result)
+        }
+      }
+    }),
+
+    isPinned: (multihash, callback) => {
+      // callback (err, pinned, reason)
```

Comment:
remove if not used

Reply:
Also, it is preferable for the callback to be use `function (err, result)` to follow the same pattern as everything else